### ISSUE TITLE
mongodb_up metric is not reflecting correctly 

### ISSUE THREAD

**Issue by ckopparthi** [percona/mongodb_exporter#347]
I am running the pmm-agent in the kubernetes cluster as a stateful, restarted the MongoDB pod and the mongodb_up is not set to zero 

pmm-agent and pmm-server version: 2.22.0

At the MongoDB pod restart the below is the log that I get
```
INFO[2021-09-27T13:16:13.057+00:00] time="2021-09-27T13:16:13Z" level=error msg="error while checking mongodb connection: server selection error: context canceled, current topology: { Type: Single, Servers: [{ Addr: mongodb-cs-cfg-0.mongodb-cs-cfg.dev.svc.cluster.local:27019, Type: Unknown, Average RTT: 0, Last error: connection() error occured during connection handshake: dial tcp: lookup mongodb-cs-cfg-0.mongodb-cs-cfg.dev.svc.cluster.local on 1: no such host }, ] }. mongo_up is set to 0"  agentID=/agent_id/3466d059-8282-48f4-8c31-dc91d17ee1c7 component=agent-process type=mongodb_exporter
INFO[2021-09-27T13:16:13.058+00:00] time="2021-09-27T13:16:13Z" level=error msg="error while checking mongodb connection: server selection error: context canceled, current topology: { Type: Single, Servers: [{ Addr: mongodb-cs-cfg-0.mongodb-cs-cfg.dev.svc.cluster.local:27019, Type: Unknown, Average RTT: 0, Last error: connection() error occured during connection handshake: dial tcp: lookup mongodb-cs-cfg-0.mongodb-cs-cfg.dev.svc.cluster.local on : no such host }, ] }. mongo_up is set to 0"  agentID=/agent_id/3466d059-8282-48f4-8c31-dc91d17ee1c7 component=agent-process type=mongodb_exporter
INFO[2021-09-27T13:16:23.058+00:00] time="2021-09-27T13:16:23Z" level=error msg="error while checking mongodb connection: server selection error: context canceled, current topology: { Type: Single, Servers: [{ Addr: mongodb-cs-cfg-0.mongodb-cs-cfg.dev.svc.cluster.local:27019, Type: Unknown, Average RTT: 0, Last error: connection() error occured during connection handshake: dial tcp: lookup mongodb-cs-cfg-0.mongodb-cs-cfg.dev.svc.cluster.local on : no such host }, ] }. mongo_up is set to 0"  agentID=/agent_id/3466d059-8282-48f4-8c31-dc91d17ee1c7 component=agent-process type=mongodb_exporter
```
Logs say that the `mongo_up is set to 0`. But it is not set. It shows as If the metrics are not scrapped.

Output of pmm-admin status

```
bash-4.2$ pmm-agent status
pmm-agent: error: unexpected status, try --help
bash-4.2$ pmm-admin status
Agent ID: /agent_id/04a8a37b-1516-4ec3-8cb8-fd43ab5b2544
Node ID : /node_id/842ea103-f249-48ff-9d26-d41be5d41a38

PMM Server:
	URL    : https://pmm-server-dev-service.monitor.svc.cluster.local:443/
	Version: 2.22.0

PMM Client:
	Connected        : true
	Time drift       : 111.474µs
	Latency          : 368.186µs
	pmm-admin version: 2.22.0
	pmm-agent version: 2.22.0
```

**Comment by denisok** [comment]
Hi @ckopparthi ,

I don't know about `mongo_up` but probably the real issue is `Addr: mongodb-cs-cfg-0.mongodb-cs-cfg.dev.svc.cluster.local:27019, Type: Unknown, Average RTT: 0, Last error: connection() error occured during connection handshake: dial tcp: lookup mongodb-cs-cfg-0.mongodb-cs-cfg.dev.svc.cluster.local on 10.232.64.10:53: no such host `

so it just can't connect. Could you please share some details on how you run it and/or maybe share more logs ?

BTW we also have https://www.percona.com/doc/kubernetes-operator-for-psmongodb/index.html  that runs with monitoring side container.
https://github.com/percona/percona-server-mongodb-operator

As well as DBaaS feature in PMM that could deploy clustered MongoDB: https://www.percona.com/doc/percona-monitoring-and-management/2.x/using/dbaas.html

Thanks,
Denys

**Comment by ckopparthi** [comment]
I restarted the ` mongodb-cs-cfg-0.mongodb-cs-cfg.dev.svc.cluster.local:27019` server. So it shouldn't be reachable. When the mongodb server is not reachable is guess `mongo_up` should be 0

**Comment by denisok** [comment]
ah, I see, so you are saying metric is not sent. Would check.

**Comment by ckopparthi** [comment]
@denisok Thanks of quick reply

**Comment by denisok** [comment]
@ckopparthi with pmm we pass compatibility flag and thus it affected by different bug. @percona-csalguero says it will be fixed by #348 so if there are no metrics it still will get you that metric,

**Comment by ckopparthi** [comment]
@denisok Thanks for the update, can you please let us know when will this fix be released. Do you have any estimated time

**Comment by denisok** [comment]
@ckopparthi let see if we can make this:
https://github.com/percona/mongodb_exporter/milestone/3

I would say before 5th Oct we should have some release.

**Comment by denisok** [comment]
https://github.com/percona/mongodb_exporter/releases/tag/v0.20.8 released

**Comment by ckopparthi** [comment]
Do you mean, it still hangs in the latest version of mongodb_exporter?

**Comment by denisok** [comment]
don't know yet

**Comment by percona-csalguero** [comment]
In the latest version, the `mongodb_up` metric is reflecting the current state.
I ran a test starting the exporter's sandbox, connected an exporter, then stopped the instance and checked the exporter's output:

```
# HELP mongodb_up Whether MongoDB is up.
# TYPE mongodb_up gauge
mongodb_up 0
```

In the exporter's log you can see the error but that doesn't breaks the exporter:
```
ERRO[0172] cannot run getDiagnosticData: server selection error: server selection timeout, current topology: { Type: Single, Servers: [{ Addr: 127.0.0.1:17001, Type: Unknown, Last error: connection() error occured during connection handshake: dial tcp 127.0.0.1:17001: connect: connection refused }, ] }
ERRO[0172] cannot decode getDiagnosticData: <nil> for data field: unexpected data type
```

**Comment by sitoc** [comment]
Maybe the dubious point is the "mongo_up is set to 0" in the exporter log, but the metric name is "mongodb_up", different name caused the exception, I think. Looking forward to your answer, thanks

**Comment by tmikulin** [comment]
also not working for me, the mr for this was https://jira.percona.com/browse/PMM-8954 but I tried versions for mongodb exporters linux versions; **0.20.8**, **0.20.9** and **0.30.0** and they don't work (I have a non container environment)! The exporters log gives: "Cannot connect to MongoDB, server slelection error, context canceled bla bla"....

This is kind of a big deal, because I installed the exporters just for that single info to see if the mongodb is running or not!

My server is Ubuntu 18, and the version of mongo running is 4.4.1

tnx,
Tom

**Comment by denisok** [comment]
@tmikulin could you please provide more detail? If mongodb_exporter couldn't connect to mongo - it couldn't report any issues, how it would know if mongo up or down. When it is connected - yes it should report the mongodb_up anyway but with 0.

In your case, does it happens on a start ? do you provide right creds?

**Comment by tmikulin** [comment]
@denisok My issue is as described it this ticket https://jira.percona.com/browse/PMM-8954, this specificaly doesn't work "**we are expecting that mongodb_exporter should respond even mongodb database stopped and it should give mongodb_up=0**"

When the mongdb stoppes working, the mongodb exporter doesn't give mongdb_up=0, but an error that it can't connect to the db, so how can I get an info if mongodb is working or not?

**Comment by denisok** [comment]
@tmikulin could you please provide logs? As you see original ticket is solved so it should work.
Please enable `--log.level=debug` and connect to the mongodb, then get mongodb down. In parallel could you please gather curl output from the metrics endpoint - when it works and when it doesn't. Also give it a little time - 5min or so to see if there some large timeouts involved.

And please provide logs and outputs of your experiment.

We have tried to reproduce this issue, but it was working on our side.

**Comment by tmikulin** [comment]
I think it's the version of the mongodb server used, but to recreate it I used these files:

```
version: '3.7'
services:
  mongodb_container:
    image: mongo:latest
    environment:
      MONGO_INITDB_ROOT_USERNAME: root
      MONGO_INITDB_ROOT_PASSWORD: rootpassword
    ports:
      - 27017:27017
    volumes:
      - mongodb_data_container:/data/db

volumes:
  mongodb_data_container:
```

and for running the mongodb exporter I used the most recent version:

`docker run -d --name mongodb_exporter -p 9216:9216 -e "MONGODB_URI=mongodb://root:rootpassword@192.168.0.19:27017" --ip=IP_ADDRESS percona/mongodb_exporter:0.30
`

Stop the mongodb server container, and the mongodb exporter just doesn't respond anymore...

tnx,
Tom

**Comment by denisok** [comment]
@tmikulin I don't remember that we added environment support to mongodb_exporter. Could you please pass `--mongodb.uri`:

`docker run -d -p 9216:9216 -p 27017:27017 --ip=IP_ADDRESS percona/mongodb_exporter:0.30 --mongodb.uri=mongodb://root:rootpassword@192.168.0.19:27017`

**Comment by tmikulin** [comment]
same thing man....the exporter stops working....

**Comment by percona-csalguero** [comment]
I am currently working on (PMM-9312](https://jira.percona.com/browse/PMM-9312) and I made the fix as part of that ticket.

**Comment by frenkye** [comment]
This issue is real problem, the exporter is not catching up when mongo doesn't run. So it fail on timeout when you try access `/metrics`. As it could be seen here:
```
:~$ curl localhost:6157
<html>
<head>
	<title>MongoDB exporter</title>
</head>
<body>
	<h1>MongoDB exporter</h1>
	<p><a href="/metrics">Metrics</a></p>
</body>
</html>

:~$ curl localhost:6157/metrics
An error has occurred while connecting to MongoDB:

cannot connect to MongoDB: server selection error: server selection timeout, current topology: { Type: Single, Servers: [{ Addr: localhost:27017, Type: Unknown, Last error: connection() error occured during connection handshake: dial tcp [::1]:27017: connect: connection refused }, ] }
```
If mongodb is up, all metrics are shown properly.

Also passing secrets as ENV is propably best solution, if you don't have config file. Passing it as --mongodb.uri parameter is how you expose secrets in process names, if you checking running processes and sending this data somewhere else.

IMO just set timeout for connection to mongodb and on timeout show `mongodb_up 0`

**Comment by sitoc** [comment]
The same issue for me, mongodb_exporter couldn't set the mongodb_up=0 when mongodb instance go down, mongodb_exporter: 0.29.0,  0.31.1, 0.31.2, 0.32.0, these version do not fix this issue, Looking forward to the solution in the next release ,thanks a lot

**Comment by denisok** [comment]
I believe this was fixed :) please check latest version. There were couple of bugs that we fixed if mongo is not connected.

**Comment by frenkye** [comment]
Still not returning `mongodb_up 0` when mongo is down and getting empty reposnse from exporter `/metrics` endpoint

Exporter version
```
:~$ /opt/prometheus/bin/mongodb_exporter --version
mongodb_exporter - MongoDB Prometheus exporter
Version: v0.34.0
Commit: 5c3358c
Build date: 2022-08-05T09:18:49Z
```

Check:
```
:~$ systemctl status mongod.service
● mongod.service - MongoDB Database Server
     Loaded: loaded (/lib/systemd/system/mongod.service; enabled; vendor preset: enabled)
     Active: inactive (dead) since Thu 2022-09-15 08:22:06 CEST; 16s ago
       Docs: https://docs.mongodb.org/manual
    Process: 898445 ExecStart=/usr/bin/mongod --config /etc/mongod.conf (code=exited, status=0/SUCCESS)
   Main PID: 898445 (code=exited, status=0/SUCCESS)
        CPU: 1.484s

:~$ curl 'localhost:6157/metrics'
curl: (52) Empty reply from server

:~$ systemctl start mongod.service

:~$ systemctl status mongod.service
● mongod.service - MongoDB Database Server
     Loaded: loaded (/lib/systemd/system/mongod.service; enabled; vendor preset: enabled)
     Active: active (running) since Thu 2022-09-15 08:22:55 CEST; 3s ago
       Docs: https://docs.mongodb.org/manual
   Main PID: 898638 (mongod)
     Memory: 162.3M
        CPU: 1.242s
     CGroup: /system.slice/mongod.service
             └─898638 /usr/bin/mongod --config /etc/mongod.conf

:~$ curl -s 'localhost:6157/metrics' | grep "_up"
# HELP mongodb_up Whether MongoDB is up.
# TYPE mongodb_up gauge
mongodb_up 1
```

**Comment by HenrikLindholm86** [comment]
I believe this panic is fixed in main by @trvrnrth as of October 6:th  (commit `fffcfe8332a5de5a51618662d3f379c288fce7b0`)
But it's not part of the latest release since the release branch were created before that.

**Comment by frenkye** [comment]
It's possible i'am not testing master, since there are release builds. Last v0.35 dont have this commit inside from 6th Oct.

**Comment by igroene** [comment]
fixed as part of https://github.com/percona/mongodb_exporter/pull/653
